presage-store-sled/src/protocol.rs

```diff
@@ -51,6 +48,38 @@ impl SledProtocolStore<PniSledStore> {
     }
 }
 
+impl<T: SledTrees> SledProtocolStore<T> {
+    fn next_key_id(&self, tree: &str) -> Result<u32, SignalProtocolError> {
+        Ok(self
+            .store
+            .db
+            .read()
+            .expect("poisoned mutex")
+            .open_tree(tree)
+            .map_err(|e| {
+                log::error!("sled error: {}", e);
+                SignalProtocolError::InvalidState("next_key_id", "sled error".into())
+            })?
+            .into_iter()
+            .keys()
+            .filter_map(Result::ok)
+            .last()
+            .and_then(|data| Some(u32::from_be_bytes(data.as_ref().try_into().ok()?)))
+            .map_or(0, |id| id + 1))
+    }
+
+    /// Whether to force a pre key refresh.
+    ///
+    /// Check whether we have:
+    /// - 1 signed EC pre key
+    /// - 1 Kyber last resort key
+    async fn needs_pre_key_refresh(&self) -> Result<bool, SignalProtocolError> {
+        let has_signed_pre_keys = self.signed_pre_keys_count().await? > 0;
+        let has_last_resort_kyber_pre_keys = self.kyber_pre_keys_count(true).await? > 0;
+        Ok(has_signed_pre_keys && has_last_resort_kyber_pre_keys)
```

Comment:
I think it is better to short circuit this query.

Reply:
what do you mean? if the left hand side is false it'll be short circuited.

Reply:
Removed since it is unused